analyzers/src/SonarAnalyzer.Common/Helpers/CountComparisonSyntax.cs

```diff
@@ -0,0 +1,118 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Linq;
+
+namespace SonarAnalyzer.Helpers
+{
+    public readonly struct CountComparisonSyntax : IEquatable<CountComparisonSyntax>
+    {
+        public int? Left { get; }
+        public int? Right { get; }
+        public ComparisonKind Operator { get; }
```

Comment:
Normalization is key component of this class. And also a key pain. Let's change the inner state to normalized one. To store
* IsValid
* ComparisonKind Operator
* int Value

`int?` should not be needed any more, as we don't care when `IsValid = false`, right?

That should make things simple.

Reply:
@pavel-mikula-sonarsource I like the suggestion storing the result of the analysis. However, that is - I think - not enough.

If we take a look to the rule `DoNotCheckZeroSizeCollection` we can - and I think, should - extend this rule to all statements that always evaluate to to `true` or `false`. So `collection.Count < -1`, and `array.Length > -10` should also trigger this one. In that case, you need the inter value and the comparison operator to give a describtive message.

We might want to consider to do the normalization directly, and only store the constant to compare with, although it might turn out that for an additional fix provider it worth knowing, which part is where. I think that is a hard call.

The idea was to get the normalization part in the class/struct only, to localize the pain. When that has been achieved (and I think we did that now), we can focus on on reducing the pain too in this struct. What do you think?

Reply:
We can store the normalized state to make this a class and remove all the code that makes it complicated. And we can also store the original state to be able to generate proper error messages.
That would be something like

// Normalized state (private)
* IsValid
* Operator (the normalized one)
* Value

// Original state (public)
* ExpressionLeft (or OriginalLeft)
* ExpressionRight (or OriginalRight)
* ExpressionOperator (or OriginalOperator)

My key idea is to achieve this:
```
public bool IsEmpty() =>
            IsMatch(ComparisonKind.Equals, 0)
            || IsMatch(ComparisonKind.LessThan, 1)
            || IsMatch(ComparisonKind.LessThanOrEqual, 0);
```
because it simple and straight forward. We don't need any additional equal-logic and persistent instances:
```
        private static readonly CountComparisonSyntax[] Empties = new[]
        {
            new CountComparisonSyntax(null, ComparisonKind.Equals, 0),
            new CountComparisonSyntax(null, ComparisonKind.LessThan, 1),
            new CountComparisonSyntax(null, ComparisonKind.LessThanOrEqual, 0),
        };
```

Reply:
I still think that you want to communicate something like this: "array.length < -1 always evaluates as false, and therfor without meaning.

And for the fix for the `Any()` and `!Any()` is it beneficial to use `CountComparisonSyntax`, and by doing so, not have to evaluate twice which part of the expression is the constant.

For me, the added value here, is to make the code that uses this, the easiest to read/reason about. And if that makes this a bit harder to read, I think that's the compromise.

Reply:
I'm sorry but I'm lost with the latest commit. Where did the `CountComparisonSyntax` type go? It was a good type for holding the complicated logic on one isolated place.

Reply:
Well, the outcome of the evalutation is now a simple enum, al the complex logic is here: https://github.com/SonarSource/sonar-dotnet/pull/4033/files#diff-517470f94b2d684730153b4a3f4011df2483f680ed478bfc323cd3aa07ea1d99